### PR TITLE
cleanup: split filtering & ordering in OptionsListUtils

### PR DESCRIPTION
**Pull request by hannojg**
### Explanation of Change

This PR is a refactoring that clearly splits the responsibilities into:

- filtering options
- ordering options

Before ordering and filtering was happening inter-connectedly, which makes it hard to change anything about that code.

Over at this PR i am working on improving the performance of the filtering:

- https://github.com/Expensify/App/pull/51954

and i need the filtering and ordering to be two clearly separate operations (so we can easily just swap out the filtering for a faster implementation).

#### Notes about the changes:

- I found that the `excludeUnknownUsers`  was a option that had no code path that was ever used, so I was able to remove it
- Likewise `sortPersonalDetailsByAlphaAsc` was default `true` and there was no code path that was setting this to `false` so I removed that as well
- For ordering the options we need to sort by the `lastVisibleActionCreated`. It was currently not present on the option and so i added it to `createOption`
- I renamed `getOptions` to `getValidOptions` to emphasis the functions responsibility to return us only the options that are valid for showing in the UI
- I created separate functions for filtering and ordering. The new function to get a filtered and ordered options object is now called `filterAndOrderOptions` (before it was `filterOptions`). The new separate functions are:
	- `orderReportOptions`
	- `orderPersonalDetailsOptions`
	- `orderOptions`, combines the two above functions
	- `filterReports`
	- `filterPersonalDetails`
	- `filterCurrentUserOption` 
	- `filterUserToInvite`
	- `filterOptions` combines the four above

There is also a performance advantage to this I want to point out:

Before we were applying some ordering to _all_ options before filtering them. Now, we filter the options first and then apply the ordering _to the remaining search result_, which can be way more efficient.

### Fixed Issues

$ https://github.com/Expensify/App/pull/51954
PROPOSAL: https://github.com/Expensify/App/pull/51954#issuecomment-2464282215

### Tests

- [ ] Verify that no errors appear in the JS console
- [ ] Open the search:
	- Confirm that search results between this branch and production are very much the same (it might be that on dev / production you've different items loaded which can slightly influence the search results, but largely they should match)

### QA Steps

Same as testing

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I used [JaimeGPT](https://chatgpt.com/g/g-2dgOQl5VM-english-to-spanish-translator-aka-jaimegpt) to get English > Spanish translation. I then posted it in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I added [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) for any new feature or bug fix in this PR to help automatically prevent regressions in this user flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

CleanShot 2024-12-02 at 14 48 07

</details>

<details>
<summary>Android: mWeb Chrome</summary>

CleanShot 2024-12-02 at 14 48 07

</details>

<details>
<summary>iOS: Native</summary>

CleanShot 2024-12-02 at 14 48 07

</details>

<details>
<summary>iOS: mWeb Safari</summary>

CleanShot 2024-12-02 at 14 48 07

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

CleanShot 2024-12-02 at 14 48 07

</details>

<details>
<summary>MacOS: Desktop</summary>

CleanShot 2024-12-02 at 14 48 07

</details>

### CODE REVIEW

**Comment by hannojg** [comment]
cc @ahmedGaber93 since you reviewed all PRs of this cleanup so far 🫶 
This is ready for review (will mark it as such once i've filled all the details)

File: tests/unit/OptionsListUtilsTest.ts
Comment (neil-marcellini) [inline tests/unit/OptionsListUtilsTest.ts:534]: Why are these commented out? Please uncomment or remove and explain why.

File: tests/unit/OptionsListUtilsTest.ts
Comment (neil-marcellini) [inline tests/unit/OptionsListUtilsTest.ts:892]: Why was this removed? Is it redundant? If so where is it covered now?

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Nice work! It mostly looks good to me. I recognize that it's pretty hard to test such a large refactor like this, but I will add a C+ to stick to our process and have another set of eyes.

I have some questions about the unit tests.

**Comment by ahmedGaber93** [comment]
Reviewing and testing today.

File: tests/unit/OptionsListUtilsTest.ts
Comment (ahmedGaber93) [inline tests/unit/OptionsListUtilsTest.ts:959]: I didn't get that why length changed from 4 to 2?

**Review by ahmedGaber93 — COMMENTED** [review]
Great refactoring, just small notes

**Comment by ahmedGaber93** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] For any bug fix or new feature in this PR, I verified that sufficient [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) are included to prevent regressions in this flow.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

File: src/libs/OptionsListUtils.ts
Comment (hannojg) [reply src/libs/OptionsListUtils.ts:983]: you would just do it online with the return statement?

File: tests/unit/OptionsListUtilsTest.ts
Comment (hannojg) [reply tests/unit/OptionsListUtilsTest.ts]: I removed this now (thanks for pointing it out!). This test case only worked when limiting the options using `maxRecentReportsToShow` (because it changes the results in the options).

With these changes `getValidOptions` doesn't offer the possibility to anymore to limit the results (as no call side in the app needs it, we only need it when filtering and ordering)

File: tests/unit/OptionsListUtilsTest.ts
Comment (hannojg) [reply tests/unit/OptionsListUtilsTest.ts:892]: Yes, this test was consolidated into this test here:

https://github.com/margelo/expensify-app-fork/blob/2eaeaf42ed7696ff3dc98b34567ee17be4cd2f5c/tests/unit/OptionsListUtilsTest.ts#L947-L957

File: tests/unit/OptionsListUtilsTest.ts
Comment (hannojg) [reply tests/unit/OptionsListUtilsTest.ts]: Good catch, thanks! There was still something broken about the order and limit part. I fixed that now, reverted the changes in this test case and now the test case is passing again!

File: src/libs/OptionsListUtils.ts
Comment (hannojg) [reply src/libs/OptionsListUtils.ts]: fixed this one! also realized i was missing one other parameter to the ordering which i added now

**Comment by hannojg** [comment]
Guys, thank you for the excellent review! 🫶 

I have everything applied from the review, but i think there is something broken with the search results. While searching the results between this PR and prod seem the same but when opening the search they look vastly different:

CleanShot 2024-12-04 at 13 59 19

Will mark as ready again once thats fixed

File: src/libs/OptionsListUtils.ts
Comment (hannojg) [reply src/libs/OptionsListUtils.ts]: actually i probably have [to rewrite this](https://github.com/Expensify/App/pull/53371/commits/e26cccbf2393660dd00a215bf7f12dadef288529) a bit. I think in the original implementation there was kind of a bug. Will explain in detail once fixed …

Comment (hannojg) [reply src/libs/OptionsListUtils.ts]: Okay, so we don't need the additional `private_isArchived`, we don't need the additional `isSelfDM` check.

The code i have written makes sure to check all of these properties only once (before we were using those options in the multiple order functions, which makes it harder to understand whats happening).

The `?? ''` is actually correct, i added a code comment to better explain it

**Comment by hannojg** [comment]
Okay, prod and dev looked the same now in my test cases:

CleanShot 2024-12-04 at 17 30 59

File: tests/unit/OptionsListUtilsTest.ts
Comment (neil-marcellini) [inline tests/unit/OptionsListUtilsTest.ts:991]: This doesn't make sense to me. We're asking it show 0 recent reports max, then we assert it shows 10. I think we should leave the value undefined if we don't want to filter at all. If it's 0, we should return 0 recent reports.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Thanks for the updates, all my requested changes from last time were addressed. It all looks good to me except for one test. Let's also please add more test steps with more detail, trying to cover most of the areas that were changed. If we do that it's less likely we'll have to revert this due to regressions on staging.

**Comment by neil-marcellini** [comment]
Another option is to ask Applause to run the regression tests on this branch, I think we do that sometimes for large changes. I'm not sure it's appropriate for this one, but we can ask if you want to try that.

**Comment by ahmedGaber93** [comment]
I tested it as much as I could. Search result look the same on dev and staging except the last case

<details>
<summary> ✅  search tab</summary>

https://github.com/user-attachments/assets/fd5998ba-6d1d-4103-9ae5-a2faf2b83ff6

</details>

<details>
<summary> ✅  search from header icon</summary>

https://github.com/user-attachments/assets/308d305e-9d4f-4602-9c6a-bb7b971c48ab

</details>


<details>
<summary> ✅  multi search terms</summary>

https://github.com/user-attachments/assets/7daa3b39-7a8a-490c-affa-dd7aa02ee6f6

</details>


<details>
<summary> ✅  self DM</summary>

https://github.com/user-attachments/assets/f106f917-d252-4ed1-86dc-c5665d7cab82



</details>


<details>
<summary> ✅  user to join - email</summary>



https://github.com/user-attachments/assets/ddaab718-2b83-44a4-9b0b-a09fcbccfd5c



</details>


<details>
<summary> ✅  user to join - phone</summary>

https://github.com/user-attachments/assets/5172ef82-9f06-4a7c-92ad-94ad60ccf913


</details>


<details>
<summary> ✅  start chat</summary>


https://github.com/user-attachments/assets/2083438c-8b58-4d66-8f4e-083d4fc7250d

</details>

<details>

<summary> ❌ create expense</summary>

1. Fresh login with a high traffic account that have a lot of workspaces on DEV and staging
2. Click Fab icon > Create expense > choose participant

The recent reports are not the same, especially "Big chat" workspace.

Screenshot 2024-12-05 at 9 46 04 PM

</details>

File: tests/unit/OptionsListUtilsTest.ts
Comment (hannojg) [reply tests/unit/OptionsListUtilsTest.ts]: yeah i get why that one is confusing, let me explain:

- before those refactorings `maxRecentReportsToShow` was by default `0`. When passing 0 it was expected to return all reports. 
- as there were more call sites passing `maxRecentReportsToShow: 0` then actually limiting them, i changed the parameter to be optional and by default return all reports. I removed all call sites using `maxRecentReportsToShow: 0`
- [In this review comment](https://github.com/Expensify/App/pull/53371#discussion_r1868351909) @ahmedGaber93 pointed out that we should add a test to verify that `maxRecentReportsToShow: 0` are still returning all reports and i added this test

Now i realized and agree with Neil that passing `0` is confusing and you'd expect to get `0` elements back in that case (as the parameter is optional now and the default is to return all items).

I extended one test case that now verifies that when passing `maxRecentReportsToShow: 0` no report results are returned here: https://github.com/Expensify/App/pull/53371/commits/dfa6fd6347d945c4a1190d3f9001901360617ab3

I hope this is fine now with everyone?

**Comment by hannojg** [comment]
@ahmedGaber93 Thanks for the thorough testing! Can you please test this case again?

**Comment by hannojg** [comment]
@neil-marcellini yeah i would sleep better if this PR would get a Applause regression test 😅  maybe we can ask them to only test search features / features that are related to where we show result lists

**Comment by ahmedGaber93** [comment]
> Can you please test this case again?

Hmm! Ordering still not the same.

File: src/pages/iou/request/MoneyRequestParticipantsSelector.tsx
Comment (ahmedGaber93) [inline src/pages/iou/request/MoneyRequestParticipantsSelector.tsx]: I think `config` will be ignored when we're passing `searchValue`  as `undefined` because `orderReportOptions(options: ReportUtils.OptionData[])` doesn't have a `config` parameter

File: src/pages/iou/request/MoneyRequestParticipantsSelector.tsx
Comment (ahmedGaber93) [reply src/pages/iou/request/MoneyRequestParticipantsSelector.tsx:131]: The ordering looks the same with me in this case only when remove `sortComparatorReportOptionByDate` [here](https://github.com/margelo/expensify-app-fork/blob/dfa6fd6347d945c4a1190d3f9001901360617ab3/src/libs/OptionsListUtils.ts#L928), but I don't test this change effect on the other cases

Comment (hannojg) [reply src/pages/iou/request/MoneyRequestParticipantsSelector.tsx:131]: Hm, okay, will double check with multiple accounts - good observation with the config not being used, thanks!

**Comment by hannojg** [comment]
@ahmedGaber93 Okay I think I fixed the ordering now. I did this by separating the order statements, this is how it was working before (ie. the first order function was called in `getOptions` the second in `filterOptions` - both functions are now called in the `orderReportOptions` functions).

I will revisit the ordering function after the cleanup PR is merged. I think sorting twice is quite inefficient and especially lodash orderBy is known to be slow (but will profile soon and provide before and after values!)

**Review by ahmedGaber93 — APPROVED** [review]
Tested well! All yours @neil-marcellini.

**Review by neil-marcellini — APPROVED** [review]
Ok, looks like it's time to merge. Hopefully QA will catch any issues we missed and if they create a blocker issue we'll revert this and fix the specific problem.

**Pull request merged**
